src/libqof/qof/gnc-numeric.hpp

```diff
@@ -0,0 +1,418 @@
+private:
+    struct round_param
+    {
+        int64_t num;
+        int64_t den;
+        int64_t rem;
+    };
+    /* Calculates the denominator required to convert to figs sigfigs. */
+    int64_t sigfigs_denom(unsigned figs) const noexcept;
+    /* Calculates a round_param struct to pass to a rounding function that will
+     * finish computing a GncNumeric with the new denominator.
+     */
+    round_param prepare_conversion(int64_t new_denom) const;
+    int64_t m_num;
+    int64_t m_den;
+};
+
+/**
+ * \defgroup gnc_numeric_arithmetic_operators
+ * @{
+ * Normal arithmetic operators. The class arithmetic operators are implemented
+ * in terms of these operators. They use GncRational operators internally then
+ * call the GncNumeric(GncRational&) constructor which will silently round
+ * half-down to obtain int64_t members.
+ *
+ * These operators can throw std::overflow_error, std::underflow_error, or
+ * std::invalid argument as indicated in the class GncNumeric documentation.
+ *
+ * \param a The right-side operand
+ * \param b The left-side operand
+ * \return A GncNumeric computed from the operation.
+ */
+GncNumeric operator+(GncNumeric a, GncNumeric b);
+inline GncNumeric operator+(GncNumeric a, int64_t b)
+{
+    return a + GncNumeric(b, 1);
+}
+inline GncNumeric operator+(int64_t a, GncNumeric b)
+{
+    return b + GncNumeric(a, 1);
+}
+GncNumeric operator-(GncNumeric a, GncNumeric b);
+inline GncNumeric operator-(GncNumeric a, int64_t b)
+{
+    return a - GncNumeric(b, 1);
+}
+inline GncNumeric operator-(int64_t a, GncNumeric b)
+{
+    return b - GncNumeric(a, 1);
+}
+GncNumeric operator*(GncNumeric a, GncNumeric b);
+inline GncNumeric operator*(GncNumeric a, int64_t b)
+{
+    return a * GncNumeric(b, 1);
+}
+inline GncNumeric operator*(int64_t a, GncNumeric b)
+{
+    return b * GncNumeric(a, 1);
+}
+GncNumeric operator/(GncNumeric a, GncNumeric b);
+inline GncNumeric operator/(GncNumeric a, int64_t b)
+{
+    return a / GncNumeric(b, 1);
+}
+inline GncNumeric operator/(int64_t a, GncNumeric b)
+{
+    return b / GncNumeric(a, 1);
+}
+/** @} */
+/**
+ * std::stream output operator. Uses standard integer operator<< so should obey
+ * locale rules. Numbers are presented as integers if the denominator is 1, as a
+ * decimal if the denominator is a multiple of 10, otherwise as
+ * "numerator/denominator".
+ */
+
+/* Implementation adapted from "The C++ Standard Library, Second Edition" by
+ * Nicolai M. Josuttis, Addison-Wesley, 2012, ISBN 978-0-321-62321-8.
+ */
+template <typename charT, typename traits>
+std::basic_ostream<charT, traits>& operator<<(std::basic_ostream<charT, traits>& s, GncNumeric n)
+{
+    std::basic_ostringstream<charT, traits> ss;
+    ss.copyfmt(s);
+    ss.width(0);
+    if (n.denom() == 1)
+         ss << n.num();
+    else if (n.denom() % 10 == 0)
+        ss << n.num() / n.denom() << "."
+           << (n.num() > 0 ? n.num() : -n.num()) % n.denom();
```

Comment:
You're right, and not only that the condition is wrong: a denom of e.g. 130 would run that code and produce an incorrect representation.